### ISSUE TITLE
crypto_password_hash/2 incompatible between 8.5.13 and 8.5.14

### ISSUE THREAD

**Issue by matko** [SWI-Prolog/packages-ssl#168]
It looks like something changed between 8.5.13 and 8.5.14 which makes `crypto_password_hash/2` behave differently. This causes hashes that were generated under 8.5.13 and earlier to not validate under 8.5.14 and later.

Example:
```
$ swivm use v8.5.13
Now using SWI-Prolog v8.5.13
$ swipl -q
?- crypto_password_hash("root", X).
X = '$pbkdf2-sha512$t=131072$fHxS+fONBql1sIqx5wUw2g$JCdZjj3ZLqgl1Yj5jRvDoCIP+FL0A13iem6eXfonBUvs0g6U6fAd8R9PTLQaqsUVawoj2S2z1lPPAMd3EfYVvw'.

?- crypto_password_hash("root", '$pbkdf2-sha512$t=131072$fHxS+fONBql1sIqx5wUw2g$JCdZjj3ZLqgl1Yj5jRvDoCIP+FL0A13iem6eXfonBUvs0g6U6fAd8R9PTLQaqsUVawoj2S2z1lPPAMd3EfYVvw').
true.

?- ^D
$ swivm use v8.5.14
Now using SWI-Prolog v8.5.14
$ swipl -q
?- crypto_password_hash("root", '$pbkdf2-sha512$t=131072$fHxS+fONBql1sIqx5wUw2g$JCdZjj3ZLqgl1Yj5jRvDoCIP+FL0A13iem6eXfonBUvs0g6U6fAd8R9PTLQaqsUVawoj2S2z1lPPAMd3EfYVvw').
false.
```

It looks like the hashes generated from 8.5.14 onwards are significantly longer than before:
```
$ swivm use v8.5.13
Now using SWI-Prolog v8.5.13
$ swipl -q
?- crypto_password_hash("root", X).
X = '$pbkdf2-sha512$t=131072$XrBtPNc8VQwAR5pIaqQXcQ$2RwAQA8a65r2ykPzJa7r0Pleh6q2uI1h2QfZ0oIqAjdpgQpHWxlDVGB/FucGXvJ0AFgTtIfkPK38Ymu+oiaD5w'.

?- ^D
$ swivm use v8.5.14
Now using SWI-Prolog v8.5.14
$ swipl -q
?- crypto_password_hash("root", X).
X = '$pbkdf2-sha512$t=131072$LwQ7w6VXesKFw4HDrsOUKiXDg8KqPSE$w7zDk8O6wpJDw7fCm8ONVhQcWl9zw5B5fXk/A1LDu8OWwqjCnz/DhDQAZXbCsMKawoJtGsKVw5zDh8OGeEfDt8KSEQlmwoI+w7sTw7ZqwoQ1w6jCr8ONwqwmw57CkxBw'.
```

As I understand the documentation of `crypto_passworrd_hash` this simply should not happen. Changes to the way `crypto_password_hash` generates new hashes should not impact its validation of old hashes.

I'm not sure what changed though. Looking at the changelog between 8.5.13 and 8.5.14 doesn't point at anything obvious.

**Comment by JanWielemaker** [comment]
Moved to ssl package.   But, the source for the ssl package is not changed between these versions.    Possibly OpenSSL changed between the two compilation runs?   Could it be related to the uninitialized structure you fixed recently?

**Comment by matko** [comment]
I considered that it might be another build environment issue, so I tested this in a fresh debian stable container, where I built both 8.5.13 and 8.5.14 without running an upgrade inbetween. So definitely same build environment, same order of execution, same everything. Same thing happens.

I have no idea what is causing it but the issue is definitely there. It doesn't look like zeroing that struct affected this. hashes generated in 8.5.16 validate fine in 8.5.15 and 8.5.14, but not in 8.5.13 or earlier.

**Comment by JanWielemaker** [comment]
Very weird.   As you've got all setup, could you trace through both versions and see which call is different?   There have been quite a few changes to the text handling of Prolog.   You'd expect that to result in some test failing though.

**Comment by matko** [comment]
Apologies, I've not had time to look into this issue. I guess this now got released in 9.0.0? If so, that is a pretty serious regression.

**Comment by JanWielemaker** [comment]
Anyway, I just found the culprit.  See 80e4458ca10578c8c3837e71765713cf6ce1260f.    Now I guess 9.0.1 should be released quickly, but I'm still looking for a rather old, but nasty bug that surfaces a bit too often.    I'll ask for reproducible cases on Discourse.
